File: Niveau 4/Structures de données et Balayages/Affectation des salles/main.py
import sys
import heapq
def main():
    nbSalle, nbRepresentation = map(int, input().split())
    events = []
    index = 0
    for line in sys.stdin:
        debut, fin = map(int, line.split())
        events.append((debut, fin, index))
        index += 1
        if index >= nbRepresentation:
            break
    events.sort()
    free_rooms = [(0, i) for i in range(1, nbSalle + 1)]
    heapq.heapify(free_rooms)
    room_assignment = [0] * nbRepresentation
    for start, end, idx in events:
        # Vérifie si une salle est disponible pour la représentation
        salle_disponible = False
        for _, (end_time, _) in enumerate(free_rooms):
            if end_time <= start:
                assigned_room = heapq.heappop(free_rooms)[1]
                heapq.heappush(free_rooms, (end, assigned_room))
                room_assignment[idx] = assigned_room
                salle_disponible = True
                break
        if not salle_disponible:
            print("NON")
            return
    print("OUI")
    #print(" ".join(map(str, room_assignment)))
    print(*room_assignment)
if __name__ == "__main__":
    main()

File: Niveau 4/Structures de données et Balayages/Affectation des salles/main.cpp
#include <iostream>
#include <algorithm>
#include <vector>
struct Representation
{
  int debut, fin, salle;
  bool operator<(Representation const &i)const
  {
    return debut < i.debut;
  }
};
bool tri(std::vector<Representation> &r, std::vector<int> &p, std::vector<int> &local, int s, int rp);
int main()
{
  std::ios_base::sync_with_stdio(0);
  int salle, rp;
  std::cin >> salle >> rp;
  
  std::vector<Representation> rep(rp);
  std::vector<int>local(rp);
  std::vector<int>piece(salle, -1);
  
  for (int i {0}; i < rp; i++)
  {
    std::cin >> rep [i].debut >> rep [i].fin;
    rep [i].salle = i;
  }
  std::sort(rep.begin(), rep.end());
  if (tri(rep, piece, local, salle, rp))
  {
    std::cout << "OUI" << '\n';
    for (int i = 0; i < rp; i++)
      std::cout << local [i] << " ";
    std::cout << '\n';
  }
  else
    std::cout << "NON" << '\n';
  return 0;
}
bool tri(std::vector<Representation> &r, std::vector<int> &p, std::vector<int> &local, int s, int rp)
{
  int i {0};
  for (; i < rp; i++)
  {
    int j {0};
    for (; j < s; j++)
    {
      if (r [i].debut >= p [j])
      {
        p [j] = r [i].fin;
        local [r [i].salle] = j + 1;
        break;
      }
    }
    if (j == s)
      return false;
  }
  return true;
}
